daemon/src/main/java/org/mvndaemon/mvnd/builder/SmartGraphBuilder.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2014 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.mvndaemon.mvnd.builder;
+
+import javax.enterprise.inject.Default;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.execution.ProjectDependencyGraph;
+import org.apache.maven.graph.DefaultGraphBuilder;
+import org.apache.maven.graph.GraphBuilder;
+import org.apache.maven.model.building.Result;
+import org.eclipse.sisu.Priority;
+
+/**
+ * A replacement for the {@link DefaultGraphBuilder} to which we still partly delegate.
+ */
+@Singleton
+@Named("graphBuilder")
+@Default
+@Priority(10)
+public class SmartGraphBuilder implements GraphBuilder {
+
+    private final DefaultGraphBuilder delegate;
+
+    @Inject
+    public SmartGraphBuilder(org.apache.maven.graph.DefaultGraphBuilder delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Result<? extends ProjectDependencyGraph> build(MavenSession session) {
+        final Result<ProjectDependencyGraph> result = delegate.build(session);
+        if (result.hasErrors()) {
+            return result;
+        } else {
+            final ProjectDependencyGraph defaultGraph = result.get();
+            final SmartProjectDependencyGraph smartGraph = SmartProjectDependencyGraph.fromMaven(defaultGraph,
+                    SmartProjectDependencyGraph.getRules(defaultGraph.getSortedProjects(), session));
+            return Result.success(smartGraph);
+        }
+    }
+}
```

Comment:
Overall, I'm a bit skeptic on the benefits of overriding the graph builder to wrap the existing graph by delegating to the standard graph builder, as we have to cast it later to actually access the correct instance.
It may be easier to simply wrap it when needed.  If we need to avoid wrapping it multiple times (not sure that's actually the case), we can always cache it in the session / execution context somewhere.
The problem is that it ties us to maven internals (which I'd rather avoid as much as possible) with no real benefits imho.

Reply:
> it ties us to maven internals

That's a valid point

> It may be easier to simply wrap it when needed. If we need to avoid wrapping it multiple times (not sure that's actually the case), 

Yes, we do. The graph is needed on two locations at least. 

> we can always cache it in the session / execution context somewhere.

Yeah, storing it somewhere in the session was my first thought. I have not found any general purpose map in the session or in any of its associated classes. Do we have any better option than system properties, ThreadLocal or a static variable?

Reply:
`MavenSession.getExecutionRequest().getData()` or `MavenSession.getRepositorySession().getData()` or `MavenSession.getRepositorySession().getCache()`.

Given we want to avoid stale informations, the former may be the better as it won't outlive the current build.

At the moment, it's only used once.  I suppose it's used in an additional call to compute the width ?